homeassistant/components/intellifire/config_flow.py

```diff
@@ -53,232 +73,186 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+        """Step to select a device from the cloud.
+
+        We can only get here if we have logged in. If there is only one device available it will be auto-configured,
+        else the user will be given a choice to pick a device.
+        """
+        errors: dict[str, str] = {}
+        LOGGER.debug(
+            f"STEP: pick_cloud_device: {user_input} - DHCP_MODE[{self._dhcp_mode}"
+        )
 
-        if user_input is not None:
-            if user_input[CONF_HOST] == MANUAL_ENTRY_STRING:
-                return await self.async_step_manual_device_entry()
+        if self._dhcp_mode or user_input is not None:
+            if self._dhcp_mode:
+                serial = self._dhcp_discovered_serial
+                LOGGER.debug(f"DHCP Mode detected for serial [{serial}]")
+            if user_input is not None:
+                serial = user_input[CONF_SERIAL]
+
+            # Run a unique ID Check prior to anything else
+            await self.async_set_unique_id(serial, raise_on_progress=False)
+            self._abort_if_unique_id_configured(updates={CONF_SERIAL: serial})
+
+            # If Serial is Good obtain fireplace and configure
+            fireplace = self.cloud_api_interface.user_data.get_data_for_serial(serial)
+            if fireplace:
+                return await self._async_create_config_entry_from_common_data(
+                    fireplace=fireplace
+                )
 
-            try:
-                return await self._async_validate_ip_and_continue(user_input[CONF_HOST])
-            except (ConnectionError, ClientConnectionError):
-                errors["base"] = "cannot_connect"
+        # Parse User Data to see if we auto-configure or prompt for selection:
+        user_data = self.cloud_api_interface.user_data
+        available_fireplaces: list[IntelliFireCommonFireplaceData] = list(
+            {
+                fp.serial: fp
+                for fp in user_data.fireplaces
+                if fp.serial not in self._configured_serials
+            }.values()
+        )
+
+        # Abort if all devices have been configured
+        if not available_fireplaces:
+            return self.async_abort(reason="all_devices_already_configured")
+
+        # If there is a single fireplace configure it
+        if len(available_fireplaces) == 1:
+            return await self._async_create_config_entry_from_common_data(
+                fireplace=available_fireplaces[0]
+            )
 
         return self.async_show_form(
-            step_id="pick_device",
+            step_id="pick_cloud_device",
             errors=errors,
             data_schema=vol.Schema(
                 {
-                    vol.Required(CONF_HOST): vol.In(
-                        [host.ip for host in self._not_configured_hosts]
-                        + [MANUAL_ENTRY_STRING]
+                    vol.Required(CONF_SERIAL): vol.In(
+                        [fp.serial for fp in available_fireplaces]
```

Comment:
can the user recognize their device by the serial easily, or would it make sense to use a `SelectOptionDict` with `serial` as value, but something more recognizeable as `label` (of course only if that exists)

Reply:
The user cannot recognize their device by serial / however - most people have only a single fireplace. The backing library _could_ be updated to add this feature, but, in practicality - I don't think it "matters" because of all the users only one user of this integration is blessed to have 3 fireplaces

Reply:
@jeeftor - If you need someone with multiple units to test, lemme know. I have two Heat N Glo Escape units I'd be happy to help out your integration with!

Reply:
Check out the hacs version. It's similar to what I'm doing here - but also has cloud support.
https://github.com/jeeftor/intellifire_hacs

I think I'm having some hacs issues still regarding the switching between cloud & local (but that not in this release)

Reply:
@jaredledvina  - and if you have multiple units then maybe I should update the backing library to store "Fireplace Name" in the common data on the back end.